services/blockstorage/internodesync/state_processing_blocks.go

```diff
@@ -84,9 +100,84 @@ func (s *processingBlocksState) processState(ctx context.Context) syncState {
 		}
 	}
 
-	if !s.factory.conduit.drainAndCheckForShutdown(ctx) {
+	if !s.conduit.drainAndCheckForShutdown(ctx) {
 		return nil
 	}
 
 	return s.factory.CreateCollectingAvailabilityResponseState()
 }
+
+func (s *processingBlocksState) validatePosChain(ctx context.Context, blocks []*protocol.BlockPairContainer, receivedSyncBlocksOrder gossipmessages.SyncBlocksOrder) error {
+	if receivedSyncBlocksOrder == gossipmessages.SYNC_BLOCKS_ORDER_RESERVED && s.syncBlocksOrder == gossipmessages.SYNC_BLOCKS_ORDER_ASCENDING {
+		return nil
+	} else if s.syncBlocksOrder != receivedSyncBlocksOrder {
+		return errors.New("received chunk with blocks order which does not match blockSync expected blocks order")
+	}
+	if receivedSyncBlocksOrder == gossipmessages.SYNC_BLOCKS_ORDER_DESCENDING {
+		firstBlock := blocks[0]
+		if nextBlock, _ := s.storage.GetBlock(firstBlock.TransactionsBlock.Header.BlockHeight() + 1); nextBlock != nil { // will verify hash pointer to block
+			// prepend
+			blocks = append([]*protocol.BlockPairContainer{nextBlock}, blocks...)
+		} else {
+			ref, err := s.management.GetCurrentReference(ctx, &services.GetCurrentReferenceInput{})
+			if err != nil {
+				s.logger.Error("management.GetCurrentReference should not return error", log.Error(err))
+				return err
+			}
+			if s.managementReferenceGrace > 0 { // check is enabled // TODO: due to testing incompatibility with current reference time and time.Now()
```

Comment:
since we only seem to chane the hashptr ... why do we care that we are not current ?
also management grace ... this check is weird because now is POST finality (20 minutes) so we check that current time + grace < now ... so this is wrong (which makes me feel it is all very smelly) ... 

i feel the grace check here may be over-kill